src/coreclr/pal/src/libunwind/include/libunwind-arm.h

```diff
@@ -269,32 +269,43 @@ typedef struct unw_tdep_context
   }
 unw_tdep_context_t;
 
-/* There is no getcontext() on ARM.  Use a stub version which only saves GP
-   registers.  FIXME: Not ideal, may not be sufficient for all libunwind
-   use cases.  Stores pc+8, which is only approximately correct, really.  */
+/* FIXME: this is a stub version which only saves GP registers.  Not ideal, but
+   may be sufficient for all libunwind use cases.
+   In thumb mode, we return directly back to thumb mode on return (with bx), to
+   avoid altering any registers after unw_resume. */
 #ifndef __thumb__
-#define unw_tdep_getcontext(uc) (({                                     \
-  unw_tdep_context_t *unw_ctx = (uc);                                   \
-  register unsigned long *unw_base __asm__ ("r0") = unw_ctx->regs;      \
-  __asm__ __volatile__ (                                                \
-    "stmia %[base], {r0-r15}"                                           \
-    "adds %[base], #64"                                                 \
-    "vstmia %[base], {d0-d15}"                                          \
-    : : [base] "r" (unw_base) : "memory");                              \
-  }), 0)
-#else /* __thumb__ */
-#define unw_tdep_getcontext(uc) (({                                     \
-  unw_tdep_context_t *unw_ctx = (uc);                                   \
-  register unsigned long *unw_base __asm__ ("r0") = unw_ctx->regs;      \
-  __asm__ __volatile__ (                                                \
-    ".align 2\nbx pc\nnop\n.code 32\n"                                  \
-    "stmia %[base], {r0-r15}\n"                                         \
+#define unw_tdep_getcontext(uc) ({					\
+  unw_tdep_context_t *unw_ctx = (uc);					\
+  register unsigned long *r0 __asm__ ("r0");				\
+  unsigned long *unw_base = unw_ctx->regs;				\
+  __asm__ __volatile__ (						\
+    "mov r0, #0\n"							\
+    "stmia %[base], {r0-r15}\n"						\
     "adds %[base], #64\n"                                               \
     "vstmia %[base], {d0-d15}\n"                                        \
-    "orr %[base], pc, #1\nbx %[base]\n"                                 \
+    : [r0] "=r" (r0) : [base] "r" (unw_base) : "memory");		\
+  (int)r0; })
+#else /* __thumb__ */
+#define unw_tdep_getcontext(uc) ({					\
+  unw_tdep_context_t *unw_ctx = (uc);					\
+  register unsigned long *r0 __asm__ ("r0");				\
+  unsigned long *unw_base = unw_ctx->regs;				\
+  __asm__ __volatile__ (						\
+    ".align 2\n"							\
+    "bx pc\n"								\
+    "nop\n"								\
+    ".code 32\n"							\
+    "mov r0, #0\n"							\
+    "stmia %[base], {r0-r14}\n"						\
```

Comment:
The vstmia somehow got lost in the thumb version.

Reply:
Ah, I thought we are not using `!__thumb__` branch?

Reply:
I would like to get my change upstream, so I'd prefer it to work for both thumb and arm.

Reply:
Sure, I will add it back. Current arm32 failures in CI look related. Do we need to modify `__thumb__` and `!__thumb__` branch to adapt to 6.0 differently somehow?

Reply:
e.g. one of many failures from https://helix.dot.net/api/2019-06-17/jobs/e65e7031-9f09-4651-8d14-ef7990a53d8f/workitems/Interop/console:
```
      /root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false PassingByOutTest.dll ''
      Gathering state for process 546 corerun
      Crashing thread 00000222 signal 0000000b
      Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.546.dmp
      Written 53391360 bytes (13035 pages) to core file
      Dump successfully written
      Expected: 100
      Actual: 139
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/root/helix/work/correlation
      > /root/helix/work/workitem/e/Interop/PInvoke/SizeParamIndex/PInvoke/PassingByOut/PassingByOutTest/PassingByOutTest.sh
      Expected: True
      Actual:   False
      Stack Trace:
           at Interop_PInvoke._SizeParamIndex_PInvoke_PassingByOut_PassingByOutTest_PassingByOutTest_._SizeParamIndex_PInvoke_PassingByOut_PassingByOutTest_PassingByOutTest_sh()
      Output:
        /root/helix/work/workitem/e/Interop/PInvoke/SizeParamIndex/PInvoke/PassingByOut/PassingByOutTest/PassingByOutTest.sh: line 379:   546 Segmentation fault      (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
        
```

Reply:
Great reverting __asm__ blocks completely in second commit fixed the test issue.